### ISSUE TITLE
Introduce Hanami::Query and Hanami::Command

### ISSUE THREAD

**Issue by jodosha** [hanami/model#293]
**Please note that this is a proposal that needs to be refined.**

**Query and Command are NOT mandatory, but an OPTIONAL tool to express domain concepts with objects**.

---

Complex database operations can make the implementation of a repository really complex. This is true when multiple database tables are involved.

Sometimes we want to express a domain concept with an object, instead of a method of a repository.

Example:

``` ruby
class UsersOnTrial
  include Hanami::Query

  def each(&blk)
    users.where { trial_expires_on > Date.today }
  end
end
```

This is **NOT** bound to any database table and to any convention. Tables should be avaliable with their name as private methods. Eg: `users` to access the homonym database table.

With this object we can do interesting things:

``` ruby
users = UsersOnTrial.new
users.each do |user|
  # send an email
end

# or

users.include?(user) # to check if the given user is on trial or not
```

A bonus point is to make them composable:

``` ruby
class UsersWithTrialExpiringInAWeek
  include Hanami::Query

  def each(&blk)
    UsersOnTrial.new { trial_expires_on <= Date.today + 7 }
  end
end
```

---

I'd love to hear your feedback.

/cc @hanami/core-team @hanami/contributors

**Comment by booch** [comment]
It's not clear where `users` is coming from in the first example. When you say "Tables should be avaliable with their name as private methods", are you suggesting that the `UsersOnTrial` in the example would define `users` as a private method? If so, show that in the example to clarify.

Would a `Command` be similar to a `Query`? (Like for creating/updating?) If not, it'd probably be better to put it in a separate proposal.

**Comment by jodosha** [comment]
@booch 

> It's not clear where users is coming from in the first example.

Sorry for the misunderstanding.

If you have 3 tables in your database:

  * `users`
  * `products`
  * `orders`

Then you'll have three methods available as private methods in any object that include `Hanami::Query` and `Hanami::Command`.

**Comment by jxxcarlson** [comment]
This seems like an excellent idea to me.  It is clear, powerful, but nevertheless not magical.

**Comment by booch** [comment]
@jodosha I'm kinda scared of the magic of the tables just appearing as private methods. If I'm reading some code like your example, I'm not going to have any easy way to figure out where those methods are coming from.

And what about conflicts with other methods I might write?

**Comment by pascalbetz** [comment]
@jodosha 
IMHO providing access to those tables automatically (by table name?, mapped name?) might lead to trouble:

* conflicts
* naming conventions/ugly names (unless there is some mapping taking place) especially for legacy DBs
* debugging (where does it come from, what else is available?)
*

**Comment by jodosha** [comment]
@booch @pascalbetz In understand the method conflicts issue. We have a similar issue for views that we solved easily. All the [exposed data](http://hanamirb.org/guides/actions/exposures/) is available as methods in views. In case of conflicts you can access the exposed value with `locals[:book]` instead of `#book`.

In case of conflicts in Command/Query we can think of a method like `relation(:books)`.
The goal here is to encourage developers to extract C/Q objects. Remember that we want to make Model less verbose possible.

---

@booch @pascalbetz What's your proposed API for this?

**Comment by pascalbetz** [comment]
@jodosha 

I'd just not add "magical" methods to the C/Q. A lookup method like ```relation(:books)``` or similar would be better IMHO. 

* You can produce better errors ("there is no relation called 'books', available relations are  X, Y, Z" vs. "NoMethodError: books" 
* It is clear what the "reserved keywords" are (```relations```) vs. a unlimited (theoretically, at least) list of relations
* When reading the Q/C code, you don't need to scroll up and down to see if a method is defined in the Q/C or has been added through magic

**Comment by hlegius** [comment]
@pascalbetz agreed :+1:

**Comment by accuser** [comment]
@booch @pascalbetz whilst I might not be correct, this is how I understood @jodosha initial comment on `users` and other dependencies being private methods: 

```
class UsersOnTrial
  include Hanami::Query

  private
  attr_accessor :users

  def initialize(users: UserRepository)
    @users = users
  end

  public
  def each(&blk)
    users.where { trial_expires_on > Date.today }
  end
end
```

No magic there. ;-)

**Comment by pascalbetz** [comment]
@accuser not sure about that because @jodosha talks about conflicts and in this comment https://github.com/hanami/model/issues/293#issuecomment-178809603 about methods that are available.

**Comment by jodosha** [comment]
@accuser @pascalbetz Yes, that DI injection wasn't in my mind, however it could be a good starting point.

What about this?

```ruby
module Hanami
  module Query
    def initialize(context: Hanami::Model::Context.new)
      @context = context
    end

    private
    def collection(name)
      @context[name]
    end
  end
end

class UsersOnTrial
  include Hanami::Query

  def each(&blk)
    collection(:users).where { trial_expires_on > Date.today }
  end
end
```

That `Hanami::Model::Context` is invented and doesn't exist yet. It's just an object that holds the references to all the collections. Please note that it responds to `#[]`, like `Hash`. This can be useful for testing purposes, where we may want to pass a fake collection:

```ruby
trial = UsersOnTrial.new(users: fake_collection)
```

Does it looks better and less magic? Thank you :smile:

**Comment by accuser** [comment]
@jodosha much better :smile:

**Comment by jodosha** [comment]
@accuser The idea of injecting a single repository is nice and eliminates the need of having `Hanami::Query` at all.

However, this approach doesn't solve the problem of referencing multiple collections (eg for SQL JOINs). Sure, you can inject many repositories, but that becomes too verbose.

**Comment by pascalbetz** [comment]
@jodosha i like the idea of injecting "a context". Are there any other dependencies that need to be injected?

**Comment by jodosha** [comment]
@pascalbetz Nope, as far I can see.

**Comment by booch** [comment]
I like the suggested explicit `collection(:users)`. (I don't care so much about the name `collection` --- you could pick any of a variety of other names.) Looks good.

**Comment by pascalbetz** [comment]
update: i started working on something. but some days of skying will give it a delay:-)

**Comment by jodosha** [comment]
Quick update: me and @pascalbetz just had a video call to touch base about the progress. He's doing a really good job. We should have a PR soon.

**Comment by accuser** [comment]
@jodosha @pascalbetz awesome! :+1:

**Comment by theCrab** [comment]
@jodosha how is this going with ROM now on full steam?

**Comment by jodosha** [comment]
@theCrab This doesn't gonna be included in v0.7.0.

**Comment by jodosha** [comment]
I'm closing this as it won't be part of Hanami 1.0. Maybe it will be included in the future, according to the upcoming discussion of the core team.
